src/stream.cc

```diff
@@ -24,140 +24,110 @@
-  stream->offset += sizeof(value8);
-}
-
-void write_u32(Stream* stream, uint32_t value, const char* desc) {
-  write_data_at(stream, stream->offset, &value, sizeof(value), PrintChars::No,
-                desc);
-  stream->offset += sizeof(value);
-}
-
-void write_u64(Stream* stream, uint64_t value, const char* desc) {
-  write_data_at(stream, stream->offset, &value, sizeof(value), PrintChars::No,
-                desc);
-  stream->offset += sizeof(value);
+  WriteData(buffer, length);
 }
 
-void write_memory_dump(Stream* stream,
-                       const void* start,
-                       size_t size,
-                       size_t offset,
-                       PrintChars print_chars,
-                       const char* prefix,
-                       const char* desc) {
+void Stream::WriteMemoryDump(const void* start,
+                             size_t size,
+                             size_t offset,
+                             const char* prefix,
+                             const char* desc,
+                             PrintChars print_chars) {
   const uint8_t* p = static_cast<const uint8_t*>(start);
   const uint8_t* end = p + size;
   while (p < end) {
     const uint8_t* line = p;
     const uint8_t* line_end = p + DUMP_OCTETS_PER_LINE;
     if (prefix)
-      writef(stream, "%s", prefix);
-    writef(stream, "%07" PRIzx ": ", reinterpret_cast<intptr_t>(p) -
-                                         reinterpret_cast<intptr_t>(start) +
-                                         offset);
+      Writef("%s", prefix);
+    Writef("%07" PRIzx ": ", reinterpret_cast<intptr_t>(p) -
+                                 reinterpret_cast<intptr_t>(start) + offset);
     while (p < line_end) {
       for (int i = 0; i < DUMP_OCTETS_PER_GROUP; ++i, ++p) {
         if (p < end) {
-          writef(stream, "%02x", *p);
+          Writef("%02x", *p);
         } else {
-          write_char(stream, ' ');
-          write_char(stream, ' ');
+          WriteChar(' ');
+          WriteChar(' ');
         }
       }
-      write_char(stream, ' ');
+      WriteChar(' ');
     }
 
     if (print_chars == PrintChars::Yes) {
-      write_char(stream, ' ');
+      WriteChar(' ');
       p = line;
       for (int i = 0; i < DUMP_OCTETS_PER_LINE && p < end; ++i, ++p)
-        write_char(stream, isprint(*p) ? *p : '.');
+        WriteChar(isprint(*p) ? *p : '.');
     }
 
     /* if there are multiple lines, only print the desc on the last one */
     if (p >= end && desc)
-      writef(stream, "  ; %s", desc);
-    write_char(stream, '\n');
+      Writef("  ; %s", desc);
+    WriteChar('\n');
   }
 }
 
+MemoryStream::MemoryStream() : Stream(&writer_) {}
+
+FileStream::FileStream(const char* filename)
+    : Stream(&writer_), writer_(filename) {}
+
+FileStream::FileStream(FILE* file) : Stream(&writer_), writer_(file) {}
+
+// static
+std::unique_ptr<FileStream> FileStream::CreateStdout() {
+  return std::unique_ptr<FileStream>(new FileStream(stdout));
```

Comment:
Yes, but it's only in C++14. I guess we should have our own version like binaryen does.  I was thinking that would be nicer to do in a separate PR.